File: app/api/logs.py
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import List, Optional
from datetime import datetime, timedelta
import os
import re
from app.core.security import get_current_user
from app.database.models import User
from app.core.borgmatic import BorgmaticInterface
import structlog

logger = structlog.get_logger()
router = APIRouter(tags=["logs"])

# Initialize Borgmatic interface
borgmatic = BorgmaticInterface()

@router.get("/")
async def get_logs(
    current_user: User = Depends(get_current_user),
    log_type: str = Query("borgmatic", description="Type of logs to retrieve"),
    lines: int = Query(100, description="Number of lines to retrieve"),
    search: Optional[str] = Query(None, description="Search term"),
    level: Optional[str] = Query(None, description="Log level filter"),
    start_time: Optional[datetime] = Query(None, description="Start time filter"),
    end_time: Optional[datetime] = Query(None, description="End time filter")
):
    """Get logs with optional filtering and search"""
    try:
        # Determine log file path based on type
        if log_type == "borgmatic":
            log_path = "/app/logs/borgmatic.log"
        elif log_type == "system":
            log_path = "/var/log/syslog"
        elif log_type == "application":
            log_path = "/app/logs/app.log"
        else:
            raise HTTPException(status_code=400, detail="Invalid log type")
        
        if not os.path.exists(log_path):
            return {
                "success": True,
                "logs": [],
                "total_lines": 0,
                "message": f"Log file {log_path} not found"
            }
        
        # Read log file
        with open(log_path, 'r') as f:
            all_lines = f.readlines()
        
        # Apply filters
        filtered_lines = []
        for line in all_lines:
            # Apply search filter
            if search and search.lower() not in line.lower():
                continue
            
            # Apply level filter
            if level and level.upper() not in line.upper():
                continue
            
            # Apply time filter (basic implementation)
            if start_time or end_time:
                try:
                    # Extract timestamp from log line (adjust pattern as needed)
                    timestamp_match = re.search(r'(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})', line)
                    if timestamp_match:
                        log_time = datetime.strptime(timestamp_match.group(1), '%Y-%m-%d %H:%M:%S')
                        if start_time and log_time < start_time:
                            continue
                        if end_time and log_time > end_time:
                            continue
                except:
                    pass  # Skip time filtering if timestamp parsing fails
            
            filtered_lines.append(line)
        
        # Get requested number of lines (from end)
        total_lines = len(filtered_lines)
        if lines > 0:
            filtered_lines = filtered_lines[-lines:]
        
        return {
            "success": True,
            "logs": filtered_lines,
            "total_lines": total_lines,
            "log_type": log_type,
            "log_path": log_path
        }
        
    except Exception as e:
        logger.error("Failed to get logs", error=str(e), log_type=log_type)
        raise HTTPException(status_code=500, detail=f"Failed to retrieve logs: {str(e)}")

@router.get("/types")
async def get_log_types(current_user: User = Depends(get_current_user)):
    """Get available log types"""
    log_types = [
        {
            "id": "borgmatic",
            "name": "Borgmatic Logs",
            "description": "Backup operation logs",
            "path": "/app/logs/borgmatic.log"
        },
        {
            "id": "application",
            "name": "Application Logs",
            "description": "Web UI application logs",
            "path": "/app/logs/app.log"
        },
        {
            "id": "system",
            "name": "System Logs",
            "description": "System and service logs",
            "path": "/var/log/syslog"
        }
    ]
    
    return {
        "success": True,
        "log_types": log_types
    }

@router.get("/stats")
async def get_log_stats(
    current_user: User = Depends(get_current_user),
    log_type: str = Query("borgmatic", description="Type of logs to analyze"),
    hours: int = Query(24, description="Hours to analyze")
):
    """Get log statistics for the specified time period"""
    try:
        # Determine log file path
        if log_type == "borgmatic":
            log_path = "/app/logs/borgmatic.log"
        elif log_type == "system":
            log_path = "/var/log/syslog"
        elif log_type == "application":
            log_path = "/app/logs/app.log"
        else:
            raise HTTPException(status_code=400, detail="Invalid log type")
        
        if not os.path.exists(log_path):
            return {
                "success": True,
                "stats": {
                    "total_entries": 0,
                    "error_count": 0,
                    "warning_count": 0,
                    "info_count": 0,
                    "success_rate": 0.0
                }
            }
        
        # Calculate time threshold
        threshold_time = datetime.now() - timedelta(hours=hours)
        
        # Read and analyze logs
        with open(log_path, 'r') as f:
            lines = f.readlines()
        
        stats = {
            "total_entries": 0,
            "error_count": 0,
            "warning_count": 0,
            "info_count": 0,
            "success_rate": 0.0
        }
        
        for line in lines:
            try:
                # Extract timestamp and check if within time range
                timestamp_match = re.search(r'(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})', line)
                if timestamp_match:
                    log_time = datetime.strptime(timestamp_match.group(1), '%Y-%m-%d %H:%M:%S')
                    if log_time >= threshold_time:
                        stats["total_entries"] += 1
                        
                        # Count by level
                        line_lower = line.lower()
                        if "error" in line_lower:
                            stats["error_count"] += 1
                        elif "warning" in line_lower:
                            stats["warning_count"] += 1
                        elif "info" in line_lower:
                            stats["info_count"] += 1
            except:
                continue
        
        # Calculate success rate (basic implementation)
        if stats["total_entries"] > 0:
            non_error_entries = stats["total_entries"] - stats["error_count"]
            stats["success_rate"] = (non_error_entries / stats["total_entries"]) * 100
        
        return {
            "success": True,
            "stats": stats,
            "time_period_hours": hours
        }
        
    except Exception as e:
        logger.error("Failed to get log stats", error=str(e), log_type=log_type)
        raise HTTPException(status_code=500, detail=f"Failed to retrieve log statistics: {str(e)}")

@router.delete("/clear")
async def clear_logs(
    current_user: User = Depends(get_current_user),
    log_type: str = Query("borgmatic", description="Type of logs to clear")
):
    """Clear logs (admin only)"""
    if not current_user.is_admin:
        raise HTTPException(status_code=403, detail="Admin access required")
    
    try:
        # Determine log file path
        if log_type == "borgmatic":
            log_path = "/app/logs/borgmatic.log"
        elif log_type == "application":
            log_path = "/app/logs/app.log"
        else:
            raise HTTPException(status_code=400, detail="Invalid log type")
        
        if os.path.exists(log_path):
            # Clear the log file
            with open(log_path, 'w') as f:
                f.write("")
            
            logger.info("Logs cleared", log_type=log_type, user=current_user.username)
            
            return {
                "success": True,
                "message": f"Logs cleared successfully: {log_type}"
            }
        else:
            return {
                "success": True,
                "message": f"Log file not found: {log_path}"
            }
            
    except Exception as e:
        logger.error("Failed to clear logs", error=str(e), log_type=log_type)
        raise HTTPException(status_code=500, detail=f"Failed to clear logs: {str(e)}") 

File: nodejs/src/routes/logs.js
const express = require('express');
const router = express.Router();
const { authenticateToken } = require('../middleware/auth');
const fs = require('fs-extra');
const path = require('path');

/**
 * Get logs with optional filtering and search
 * GET /api/logs
 */
router.get('/', authenticateToken, async (req, res) => {
    try {
        const { 
            log_type = 'borgmatic', 
            lines = 100, 
            search, 
            level, 
            start_time, 
            end_time 
        } = req.query;

        // Parse numeric parameters
        const linesCount = parseInt(lines, 10) || 100;

        // Determine log file path based on type
        let logPath;
        switch (log_type) {
            case 'borgmatic':
                logPath = '/app/logs/borgmatic.log';
                break;
            case 'system':
                logPath = '/var/log/syslog';
                break;
            case 'application':
                logPath = '/app/logs/app.log';
                break;
            default:
                return res.status(400).json({
                    success: false,
                    detail: 'Invalid log type'
                });
        }

        // Check if log file exists
        if (!await fs.pathExists(logPath)) {
            return res.json({
                success: true,
                logs: [],
                total_lines: 0,
                message: `Log file ${logPath} not found`,
                log_type,
                log_path: logPath
            });
        }

        // Read log file
        const logContent = await fs.readFile(logPath, 'utf8');
        const allLines = logContent.split('\n').filter(line => line.trim());

        // Apply filters
        const filteredLines = [];
        for (const line of allLines) {
            // Apply search filter
            if (search && !line.toLowerCase().includes(search.toLowerCase())) {
                continue;
            }

            // Apply level filter
            if (level && !line.toUpperCase().includes(level.toUpperCase())) {
                continue;
            }

            // Apply time filter (basic implementation)
            if (start_time || end_time) {
                try {
                    // Extract timestamp from log line (adjust pattern as needed)
                    const timestampMatch = line.match(/(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})/);
                    if (timestampMatch) {
                        const logTime = new Date(timestampMatch[1]);
                        if (start_time && logTime < new Date(start_time)) {
                            continue;
                        }
                        if (end_time && logTime > new Date(end_time)) {
                            continue;
                        }
                    }
                } catch (error) {
                    // Skip time filtering if timestamp parsing fails
                    console.warn('Failed to parse timestamp:', error.message);
                }
            }

            filteredLines.push(line);
        }

        // Get requested number of lines (from end)
        const totalLines = filteredLines.length;
        const resultLines = linesCount > 0 ? filteredLines.slice(-linesCount) : filteredLines;

        res.json({
            success: true,
            logs: resultLines,
            total_lines: totalLines,
            log_type,
            log_path: logPath
        });

    } catch (error) {
        console.error('Failed to get logs:', error.message);
        res.status(500).json({
            success: false,
            detail: `Failed to retrieve logs: ${error.message}`
        });
    }
});

/**
 * Get available log types
 * GET /api/logs/types
 */
router.get('/types', authenticateToken, async (req, res) => {
    try {
        const logTypes = [
            {
                id: 'borgmatic',
                name: 'Borgmatic Logs',
                description: 'Backup operation logs',
                path: '/app/logs/borgmatic.log'
            },
            {
                id: 'application',
                name: 'Application Logs',
                description: 'Web UI application logs',
                path: '/app/logs/app.log'
            },
            {
                id: 'system',
                name: 'System Logs',
                description: 'System and service logs',
                path: '/var/log/syslog'
            }
        ];

        res.json({
            success: true,
            log_types: logTypes
        });
    } catch (error) {
        console.error('Failed to get log types:', error.message);
        res.status(500).json({
            success: false,
            detail: 'Failed to retrieve log types'
        });
    }
});

/**
 * Get log statistics for the specified time period
 * GET /api/logs/stats
 */
router.get('/stats', authenticateToken, async (req, res) => {
    try {
        const { log_type = 'borgmatic', hours = 24 } = req.query;

        // Parse numeric parameters
        const hoursCount = parseInt(hours, 10) || 24;

        // Determine log file path
        let logPath;
        switch (log_type) {
            case 'borgmatic':
                logPath = '/app/logs/borgmatic.log';
                break;
            case 'system':
                logPath = '/var/log/syslog';
                break;
            case 'application':
                logPath = '/app/logs/app.log';
                break;
            default:
                return res.status(400).json({
                    success: false,
                    detail: 'Invalid log type'
                });
        }

        if (!await fs.pathExists(logPath)) {
            return res.json({
                success: true,
                stats: {
                    total_entries: 0,
                    error_count: 0,
                    warning_count: 0,
                    info_count: 0,
                    success_rate: 0.0
                },
                time_period_hours: hours
            });
        }

        // Calculate time threshold
        const thresholdTime = new Date(Date.now() - (hoursCount * 60 * 60 * 1000));

        // Read and analyze logs
        const logContent = await fs.readFile(logPath, 'utf8');
        const lines = logContent.split('\n').filter(line => line.trim());

        const stats = {
            total_entries: 0,
            error_count: 0,
            warning_count: 0,
            info_count: 0,
            success_rate: 0.0
        };

        for (const line of lines) {
            try {
                // Extract timestamp and check if within time range
                const timestampMatch = line.match(/(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})/);
                if (timestampMatch) {
                    const logTime = new Date(timestampMatch[1]);
                    if (logTime >= thresholdTime) {
                        stats.total_entries += 1;

                        // Count by level
                        const lineLower = line.toLowerCase();
                        if (lineLower.includes('error')) {
                            stats.error_count += 1;
                        } else if (lineLower.includes('warning')) {
                            stats.warning_count += 1;
                        } else if (lineLower.includes('info')) {
                            stats.info_count += 1;
                        }
                    }
                }
            } catch (error) {
                // Skip lines that can't be parsed
                continue;
            }
        }

        // Calculate success rate (basic implementation)
        if (stats.total_entries > 0) {
            const nonErrorEntries = stats.total_entries - stats.error_count;
            stats.success_rate = (nonErrorEntries / stats.total_entries) * 100;
        }

        res.json({
            success: true,
            stats,
            time_period_hours: hoursCount
        });

    } catch (error) {
        console.error('Failed to get log stats:', error.message);
        res.status(500).json({
            success: false,
            detail: `Failed to retrieve log statistics: ${error.message}`
        });
    }
});

/**
 * Clear logs (admin only)
 * DELETE /api/logs/clear
 */
router.delete('/clear', authenticateToken, async (req, res) => {
    try {
        // Check if user is admin
        if (!req.user.is_admin) {
            return res.status(403).json({
                success: false,
                detail: 'Admin access required'
            });
        }

        const { log_type = 'borgmatic' } = req.query;

        // Determine log file path
        let logPath;
        switch (log_type) {
            case 'borgmatic':
                logPath = '/app/logs/borgmatic.log';
                break;
            case 'application':
                logPath = '/app/logs/app.log';
                break;
            default:
                return res.status(400).json({
                    success: false,
                    detail: 'Invalid log type'
                });
        }

        if (await fs.pathExists(logPath)) {
            // Clear the log file
            await fs.writeFile(logPath, '');
            
            console.log(`Logs cleared: ${log_type} by user ${req.user.username}`);
            
            res.json({
                success: true,
                message: `Logs cleared successfully: ${log_type}`
            });
        } else {
            res.json({
                success: true,
                message: `Log file not found: ${logPath}`
            });
        }

    } catch (error) {
        console.error('Failed to clear logs:', error.message);
        res.status(500).json({
            success: false,
            detail: `Failed to clear logs: ${error.message}`
        });
    }
});

/**
 * Get real-time log stream (WebSocket-like functionality)
 * GET /api/logs/stream
 */
router.get('/stream', authenticateToken, async (req, res) => {
    try {
        const { log_type = 'borgmatic' } = req.query;

        // Determine log file path
        let logPath;
        switch (log_type) {
            case 'borgmatic':
                logPath = '/app/logs/borgmatic.log';
                break;
            case 'application':
                logPath = '/app/logs/app.log';
                break;
            case 'system':
                logPath = '/var/log/syslog';
                break;
            default:
                return res.status(400).json({
                    success: false,
                    detail: 'Invalid log type'
                });
        }

        // Set up Server-Sent Events
        res.writeHead(200, {
            'Content-Type': 'text/event-stream',
            'Cache-Control': 'no-cache',
            'Connection': 'keep-alive',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'Cache-Control'
        });

        // Send initial data
        if (await fs.pathExists(logPath)) {
            const logContent = await fs.readFile(logPath, 'utf8');
            const lines = logContent.split('\n').filter(line => line.trim());
            const recentLines = lines.slice(-50); // Last 50 lines
            
            try {
                res.write(`data: ${JSON.stringify({
                    type: 'initial',
                    logs: recentLines,
                    timestamp: new Date().toISOString()
                })}\n\n`);
            } catch (error) {
                console.error('Error sending initial data:', error.message);
                return;
            }
        }

        // Set up file watching (simplified approach)
        let lastSize = 0;
        if (await fs.pathExists(logPath)) {
            const stats = await fs.stat(logPath);
            lastSize = stats.size;
        }

        const checkForUpdates = async () => {
            try {
                if (await fs.pathExists(logPath)) {
                    const stats = await fs.stat(logPath);
                    if (stats.size > lastSize) {
                        // Read only new content efficiently
                        const newContent = await fs.readFile(logPath, 'utf8');
                        const allLines = newContent.split('\n').filter(line => line.trim());
                        const newLines = allLines.slice(-10); // Last 10 lines
                        
                        try {
                            res.write(`data: ${JSON.stringify({
                                type: 'update',
                                logs: newLines,
                                timestamp: new Date().toISOString()
                            })}\n\n`);
                        } catch (error) {
                            console.error('Error sending log update:', error.message);
                            clearInterval(interval);
                            clearInterval(keepAlive);
                            return;
                        }
                        
                        lastSize = stats.size;
                    }
                }
            } catch (error) {
                console.error('Error checking log updates:', error.message);
            }
        };

        // Check for updates every 2 seconds
        const interval = setInterval(checkForUpdates, 2000);

        // Send keep-alive every 30 seconds
        const keepAlive = setInterval(() => {
            try {
                res.write(`data: ${JSON.stringify({
                    type: 'keepalive',
                    timestamp: new Date().toISOString()
                })}\n\n`);
            } catch (error) {
                console.error('Error sending keep-alive:', error.message);
                clearInterval(interval);
                clearInterval(keepAlive);
            }
        }, 30000);

        // Clean up on client disconnect
        req.on('close', () => {
            clearInterval(interval);
            clearInterval(keepAlive);
        });

    } catch (error) {
        console.error('Failed to stream logs:', error.message);
        res.status(500).json({
            success: false,
            detail: `Failed to stream logs: ${error.message}`
        });
    }
});

/**
 * Download log file
 * GET /api/logs/download
 */
router.get('/download', authenticateToken, async (req, res) => {
    try {
        const { log_type = 'borgmatic' } = req.query;

        // Determine log file path
        let logPath;
        let fileName;
        switch (log_type) {
            case 'borgmatic':
                logPath = '/app/logs/borgmatic.log';
                fileName = 'borgmatic.log';
                break;
            case 'application':
                logPath = '/app/logs/app.log';
                fileName = 'app.log';
                break;
            case 'system':
                logPath = '/var/log/syslog';
                fileName = 'syslog.log';
                break;
            default:
                return res.status(400).json({
                    success: false,
                    detail: 'Invalid log type'
                });
        }

        if (!await fs.pathExists(logPath)) {
            return res.status(404).json({
                success: false,
                detail: 'Log file not found'
            });
        }

        // Set headers for file download
        res.setHeader('Content-Type', 'text/plain');
        res.setHeader('Content-Disposition', `attachment; filename="${fileName}"`);

        // Stream the file
        const fileStream = fs.createReadStream(logPath);
        fileStream.pipe(res);

        fileStream.on('error', (error) => {
            console.error('Error streaming log file:', error.message);
            if (!res.headersSent) {
                res.status(500).json({
                    success: false,
                    detail: 'Failed to download log file'
                });
            }
        });

    } catch (error) {
        console.error('Failed to download logs:', error.message);
        res.status(500).json({
            success: false,
            detail: `Failed to download logs: ${error.message}`
        });
    }
});

module.exports = router;
